Added Authorization support for someip2val

int0x27 opened the pull request:
resolves #68 

- Added target actuator subscriber using kuksa common GRPC API.
- JWT token authorization is implemented in `CollectorClient`.
- `cert/` contains someip2val specific JWT token and readme how to get required keys from KUKSA.val.
- Added `--help` argument for `someip_feeder` showing options and used environment variables.
- Added common `simple_log` for broker_feeder classes.
- Added env `WIPER_STATUS` to mute `###|WiperEvent|` status lines (or make them print on the same line).
- Fixed vsomeip tag to `3.1.20.3` as `master` is no longer compatible.

int0x27 committed:
Added Authorization support for someip2val

- Added target actuator subscriber using kuksa GRPC API (old streaming api
- cert/ dir contains copies of kuksa.val/cert tools / keys + someip2val specific token.
- JWT token authorization is implemented in CollectorClient
- Added --help argument for someip_feeder showing options and used environment variables.
- Added common simple_log for broker_feeder classes.
- Added env WIPER_STATUS to mute ###|WiperEvent| status lines (or make them print on the same line)

Signed-off-by: Ivan Kirchev <Ivan.Kirchev@bosch.com>

int0x27 committed:
Removed kuksa.val keys

Signed-off-by: Ivan Kirchev <Ivan.Kirchev@bosch.com>

int0x27 commented:
Need to sync proto files with kuksa.val:

https://github.com/boschglobal/kuksa.val.feeders/blob/db45fbfdfb0b1ac03be2094eb96ed9a2e13dc03a/someip2val/proto/kuksa_data_broker/sdv/databroker/v1/collector.proto#L61

int0x27 committed:
Synced protofiles from kuksa.val master

- alternative solution for patching vsomeip, don't fail if already patched

Signed-off-by: Ivan Kirchev <Ivan.Kirchev@bosch.com>

erikbosch commented on .github/workflows/someip2val_build.yml:
```diff
@@ -56,7 +52,7 @@ jobs:
               g++-aarch64-linux-gnu binutils-aarch64-linux-gnu \
               jq python3 python3-pip python3-setuptools &&
-          pip3 install -U conan
+          pip3 install conan==1.55.0
```
Would it be an idea to add a comment on why we want to pin conan.
I see a risk that sooner or later we will get a vulnerability warning and then it is good to know why we pinned it.

Is it only because newer conan does not support "info"?

int0x27 replied:
2.0 broke arm64 cross compiling and also introduce a lot of new recipes, not sure if the state is good enough, but I also fixed the revisions of some important dependencies that only work with specific conan version. It's better to get a security warning than to find some day we are not building at all

erikbosch commented on someip2val/src/lib/broker_feeder/collector_client.cc:
```diff
@@ -11,55 +11,81 @@
     metadata_ = getGrpcMetadata();
     channel_ = grpc::CreateChannel(broker_addr, grpc::InsecureChannelCredentials());
     stub_ = sdv::databroker::v1::Collector::NewStub(channel_);
+    kuksa_stub_ = kuksa::val::v1::VAL::NewStub(channel_);
```
How is the design? Do we use both APIs (sdv::databroker::v1 and kuksa::val::v1) in parallel? What is the benefit of that, or why is it needed?

int0x27 replied:
kuksa:: proto was introduced just for the target actuator subscribe as it didn't get merged in sdv:: proto.

sdv:: interfaces support streaming updates, that is lacking in new kuksa:: proto.
But currently we just update datapoints, not stream them anyway.

erikbosch commented on someip2val/proto/kuksa_data_broker/kuksa/val/v1/types.proto:
```diff
@@ -0,0 +1,286 @@
+/********************************************************************************
```
From a long term perspective I do not see it as optimal that we copy proto files so they exist in multiple repositories (but ok for now). But what would be a better approach? Having the proto-files in separate repository so that both KUKSA.val and kuksa.val.feeders can include it as a submodule, or by some other fetch-mechanism? Or a setup script that feteches the right version from kuksa.val. Any ideas @SebastianSchildt?

int0x27 replied:
protos are all over the place, some with symlinks, I may try fetching them with CMake, but not in this PR

erikbosch commented on someip2val/proto/kuksa_data_broker/kuksa/val/v1/types.proto:
```diff
@@ -0,0 +1,286 @@
+/********************************************************************************
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
```
We do not have a NOTICE file in kuksa.val.feeders, right?

int0x27 replied:
We have the same NOTICE reference in every header, better fix it in separate PR

erikbosch commented on someip2val/src/lib/broker_feeder/actuator_subscriber.cc:
```diff
@@ -0,0 +1,173 @@
+            }
+        }
+        {
+            // TODO: lock mutex
```
When do we need to fix the TODO. Consequences if we do not fix it?

int0x27 replied:
it was actually done :) I'll remove the comment

erikbosch commented on someip2val/src/lib/broker_feeder/CMakeLists.txt:
```diff
@@ -35,10 +35,19 @@ find_program(GRPC_CPP_PLUGIN_PROGRAM_ENV
   STATIC
     data_broker_feeder.cc
     collector_client.cc
+    # logger.cc
```
Do we plan to add a logger.cc? Or shall we just remove this comment?

int0x27 replied:
I'll remove it, it turned out more complex to do than expected.

erikbosch commented on someip2val/src/lib/broker_feeder/actuator_subscriber.cc:
```diff
@@ -0,0 +1,173 @@
+    actuators_()
+{
+    LOGGER_SET_LEVEL_ENV("KUKSA_DEBUG", LEVEL_INF);
+    // subscriber_context_ = client_->createClientContext();
```
Remove commented code unless we see a need to keep. If so maybe add a comment on why we keep it.

int0x27 committed:
Fixed review issues

- Fixed concurrent access to ClientContext (valgrind issues on shutdown)
- minor logging fixes

Signed-off-by: Ivan Kirchev <Ivan.Kirchev@bosch.com>

lukasmittag reviewed (approved):
What I did:
* Ran everything locally with databroker in debug mode and saw permissions 
```
2023-05-23T14:36:36.867608Z DEBUG databroker::grpc::kuksa_val_v1::val: permissions=Permissions { expires_at: None, read: Everything, actuate: Everything, provide: Everything, create: Everything }
```
* Saw Datapoints get set after executing example

int0x27 commented on someip2val/src/lib/broker_feeder/collector_client.cc:
```diff
@@ -11,55 +11,81 @@
+        auto header = "Bearer " + databroker_token;
+        grpc_metadata["authorization"] = header;
+        LOG_TRACE << "Adding authorization: " << header << std::endl;
+    }
```
Could be logged, but definitely not here: this is called very frequently
There is a --help to show you how to specify a token.
A warning can be logged upon startup

SebastianSchildt reviewed (approved):
Let's get this merged, and potentially improve smaller things in later PRs 🚗
